##Dose measures semantic similarities among DO terms and gene products. Enrichment analyses including hypergeometric model and gene set enrichment analysis are also implemented for discovering disease associations##
library(DOSE)  
##clusterProfilers implements methtods to analyze and visualize functional profiles (GO and KEGG) of gene and gene clusters##
library(clusterProfiler)
##A system for declaratively creating graphics; found in tidyverse## 
library(ggplot2)
##collection of R packages for data science## 
library(tidyverse)
##reads rectangular data (like csv, tsv, and fwf; found in tidyverse##
library(readr)
##annotating ChIP-seq data analysis##
library(ChIPseeker)
##assesses association between genomic region sets and other genomic features##
library(regioneR)
##annotation database/data/ packages for hg38 generated by UCSC##
##consists of a set of annotations that describes the genomic locations of the known genes, transcripts, exons, and CDS, for humans##
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
##a pre-made TxDb objects wrapped into the hg38 annotation data packages##
txtdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
##used for creating GRanges objects, which represent a collection of genomic ranges that each have a single start and end location on the genome##
library(GenomicRanges)
#library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

##Set working directory## 
setwd("~/Desktop/PhD_Research/WGBS_MethPipe_Islet_Cells/WGBS_R")

##Load in bed files, which have been converted to .txt files## 
#use col_names = FALSE to tell read_tsv() to not treat the first row as headings, and instead label them sequentially from X1 to Xn 
beta_cs <- read_tsv("CT_B_cellspecific.txt", col_names = FALSE)
#add column names
colnames(beta_cs)<-c("Chr","Start","End","Length")
#load tidyverse again to call select function from dplyer 
#use mutate() to add new variables 
beta_cs <- beta_cs %>%
  dplyr::select("Chr","Start","End") %>%
  mutate("ID"=row_number(), "Strand"=".", "OtherCol"="1")

alpha_cs <- read_tsv("CT_A_cellspecific.txt", col_names = F)
colnames(alpha_cs)<-c("Chr","Start","End","Length")
alpha_cs <- alpha_cs %>% 
  dplyr::select("Chr","Start","End") %>%
  mutate("ID"=row_number(), "Strand"=".", "OtherCol"="1")

##Converts UCSB BED format to GRanges object, which is used to store the location of genomic features##
#genomic coordinates (seqnames, ranges, and strand) are located on the left-hand side, and the metadata columns are located on the right-hand side 
#almost anything can be stored in the metadata portion, such as score and QC, but here we have dummy variables 
#'*' value indicates 'unknown strand' or 'unstranded'; strands can also be + or - if that value is known
beta_cs_granges <- toGRanges(as.data.frame(beta_cs))
alpha_cs_granges <- toGRanges(as.data.frame(alpha_cs))

#GRanges operates within a universe of sequences (chromosomes/contigs) and their lengths 
#seqinfo object contains info about our set of genomic features, including sequence names, sequence lengths, circularity flag, and genome assembly (e.g., hg38)
genome(beta_cs_granges) <- "hg38" #associate a genome 
genome(alpha_cs_granges) <- "hg38"
beta_cs_granges
seqinfo(beta_cs_granges)

#gaps() returns the stretches of the genome not covered by the GRanges object 
gaps(beta_cs_granges)
gaps(alpha_cs_granges)

##Having fun with GRanges functions## 
#extracts seqnames or names of the sequences in beta_cs_granges; these names must be non-NA, non-empty and unique
seqnames(beta_cs_granges)
seqlevels(beta_cs_granges) #name as seqnames
#extracts genomic ranges of HMRs (i.e., start and end position of HMR as well as HMR length)
ranges(beta_cs_granges)
#extracts total number of HMRs in GRanges object (e.g., 14522 beta-specific HMRs)
strand(beta_cs_granges)
#extracts genomic ranges without corresponding dummy variables
granges(beta_cs_granges)

# Combine into a list
list_of_regions <- list(Alpha=alpha_cs_granges,
                        Betacell=beta_cell_cs_granges)


# Apply AnnotatePeak to this list 
# GENE ASSIGNMENT
peakAnnoList <- lapply(list_of_regions, annotatePeak, TxDb=txdb,
                       tssRegion=c(-1000, 2000), verbose=FALSE)




# # Gene segment annotation
# # plotAnnoPie(peakAnno)
# # lapply(peakAnnoList, plotAnnoPie)
# plotAnnoBar(peakAnnoList)

# # Distance to TSs
# lapply(peakAnnoList, plotDistToTSS, title="Distance to Nearest TSS")
# plotDistToTSS(peakAnnoList, title="Distance to Nearest TSS")

############################


# Gene prep 
genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)


# Venn Plot of genes
# vennplot(genes_5_noAdr)
# vennplot(genes_clust_sh_5_noAdr)

############### ENRICHMENT ANALYSES
# Don't know if you'll need this following line 
names(genes) = sub("_", "\n", names(genes))

compDO <- compareCluster(geneCluster   = genes,
                         fun           = "enrichDO", #enrichGO #enrichKEGG
                         pvalueCutoff  = 0.2,
                         pAdjustMethod = "BH")

dotplot(compDO, showCategory = 10, title = "DO Enrichment Analysis")

compDO_point1 <- compareCluster(geneCluster   = genes,
                                fun           = "enrichDO",
                                pvalueCutoff  = 0.1,
                                pAdjustMethod = "BH")
dotplot(compDO_point1, showCategory = 10, title = "DO Enrichment Analysis")


###### GO 
compGO <- compareCluster(geneCluster   = genes,
                         fun = "enrichGO",
                         OrgDb = org.Hs.eg.db,
                         ont = "BP",
                         pAdjustMethod = "BH",
                         pvalueCutoff  = 0.01,
                         qvalueCutoff  = 0.05)
dotplot(compGO, showCategory = 7, title = "GO Enrichment Analysis")


######### KEGG

compKEGG_point1 <- compareCluster(geneCluster   = genes,
                                  fun           = "enrichKEGG",
                                  pvalueCutoff  = 0.1,
                                  pAdjustMethod = "BH")
dotplot(compKEGG_point1, showCategory = 10, title = "KEGG Enrichment Analysis")

compKEGG_point05 <- compareCluster(geneCluster   = genes,
                                   fun           = "enrichKEGG",
                                   pvalueCutoff  = 0.05,
                                   pAdjustMethod = "BH")
dotplot(compKEGG_point05, showCategory = 10, title = "KEGG Enrichment Analysis")

compKEGG_point01 <- compareCluster(geneCluster   = genes,
                                   fun           = "enrichKEGG",
                                   pvalueCutoff  = 0.01,
                                   pAdjustMethod = "BH")
dotplot(compKEGG_point01, showCategory = 10, title = "KEGG Enrichment Analysis")


# Wanted to know if the kidney overlap with Liver
# Was powered by liver-kidney genes identified by the Human Protein Atlas database
# Wanna look at the liver genes
as.data.frame(peakAnnoList$Liver)$geneId

Liver_gene_symbols <- unique(bitr(as.data.frame(peakAnnoList$Liver)$geneId, 
                                  fromType="ENTREZID",
                                  toType="SYMBOL",
                                  OrgDb = org.Hs.eg.db))


####### TESTs


# Try out ChipSeeker to annotate the nearest gene
# test_bed <- annotatePeak(Bcell_clust_cs, tssRegion=c(-1000,2000),
#                      TxDb=txdb, annoDb="org.Hs.eg.db")
# test <- annotatePeak(Bcell_clust_cs_granges, tssRegion=c(-1000,2000),
#              TxDb=txdb, annoDb="org.Hs.eg.db")
# 
# test_geneIDs <- as.data.frame(test)$geneId
# names(test_geneIDs) = sub("_", "\n", names(test_geneIDs))
# 
# testDO <- enrichDO(test_geneIDs,
#                    ont = "DO", 
#                    pvalueCutoff = 0.2, 
#                    pAdjustMethod = "BH",  
#                    minGSSize = 2, 
#                    qvalueCutoff = 0.2)
# dotplot(testDO, showCategory=15)
# 
# testGO <- enrichGO(test_geneIDs,
#                    ont = "BP", 
#                    pvalueCutoff = 0.2, 
#                    pAdjustMethod = "BH",  
#                    minGSSize = 2, 
#                    qvalueCutoff = 0.2,
#                    org.Hs.eg.db)
# 
# dotplot(testGO, showCategory=15)